### ISSUE TITLE
Replica becomes ready before being fully bootstrapped

### ISSUE THREAD

**Issue by Antaxify** [CrunchyData/postgres-operator#3952]
Dear Community!

Raising this as a questions as I'm not sure what exactly to expect. I have the following test scenario:

1) Create a PostgreSQL cluster with the spec below.
2) Wait for the inital backup to complete
3) Use the scripts below to generate ~20GB of random data
4) Add a replica.

What I expected:
The read replica fully bootstraps from the primary before the pod becomes "ready" and accepts connections.

What is happening:
The pod gets into "ready" state and Postgres is accepting connections although it's not fully bootstrapped. In my tests, it synced about 5GB of the 20GB when it started accepting connections. It then continues to sync (I see disk space increasing...) for ~10 minutes more.

Questions:
1) Is this intended? Shouldn't the read-replica fully synchronise before reporting ready?
2) If this is intended, is there any way to find out if it's fully sychronised?
3) Is there maybe some timeout I can increase to give it more time to synchronise?

```
apiVersion: postgres-operator.crunchydata.com/v1beta1
kind: PostgresCluster
metadata:
  annotations:
  finalizers:
    - postgres-operator.crunchydata.com/finalizer
  name: db-sgrtest10
  namespace: sgr-test
spec:
  backups:
    pgbackrest:
      configuration:
        - secret:
            name: postgres-gcs-creds
      global:
        archive-async: 'y'
        archive-push-queue-max: 100GiB
        process-max: '8'
        archive-timeout: '900'
        log-level-console: info
        log-level-file: info
        repo1-gcs-key: /etc/pgbackrest/conf.d/gcs-key.json
        repo1-path: <some-repo-path>
        repo1-retention-full: '30'
        repo1-retention-full-type: time
        spool-path: /pgdata/pg13_pgbackrest_spool
      image: <retagged-image>
      manual:
        options:
          - '--type=full'
        repoName: repo1
      metadata:
        annotations:
          cluster-autoscaler.kubernetes.io/safe-to-evict: 'false'
          safe-to-evict: 'false'
      repos:
        - gcs:
            bucket: <some-bucket>
          name: repo1
          schedules:
            full: 0 12 * * 3
  image: <retagged-image>
  instances:
    - affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: real_persistent
                    operator: In
                    values:
                      - 'true'
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    postgres-operator.crunchydata.com/cluster: db-sgrtest10
                    postgres-operator.crunchydata.com/instance-set: instance1
                topologyKey: kubernetes.io/hostname
              weight: 1
      dataVolumeClaimSpec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 160Gi
        storageClassName: fast-durable
      metadata:
        labels:
          monitoring: crunchy-postgres
          prometheus: kube-prometheus
      name: instance1
      replicas: 1
      resources:
        limits:
          cpu: 4000m
          memory: 8192Mi
        requests:
          cpu: 1000m
          memory: 8192Mi
  metadata:
    labels:
      globalTenant: 'false'
      pg-cluster: db-sgrtest10
      pg-database: db-sgrtest10
      vendor: crunchydata
      tenant: sgrtest10
      tenantId: sgrtest10
  monitoring:
    pgmonitor:
      exporter:
        image: <retagged-image>
  patroni:
    dynamicConfiguration:
      synchronous_mode: true
      synchronous_mode_strict: false
      postgresql:
        parameters:
          checkpoint_completion_target: 0.9
          default_statistics_target: 100
          effective_cache_size: 6144MB
          effective_io_concurrency: 300
          maintenance_work_mem: 512MB
          max_connections: '600'
          max_parallel_workers: 4
          max_parallel_workers_per_gather: 2
          max_wal_size: 3GB
          max_worker_processes: 4
          min_wal_size: 1GB
          random_page_cost: 1.1
          shared_buffers: 2048MB
          synchronous_commit: "on"
          wal_buffers: 16MB
          work_mem: 32MB
    leaderLeaseDurationSeconds: 30
    port: 8008
    syncPeriodSeconds: 10
  port: 5432
  postgresVersion: 13
  users:
    - databases:
        - userdb
      name: user
```

```
-- Create tables to store random data
CREATE TABLE IF NOT EXISTS random_data1 (
    id SERIAL PRIMARY KEY,
    random_text1 TEXT,
    random_text2 TEXT,
    random_text3 TEXT,
    random_text4 TEXT,
    random_text5 TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

CREATE TABLE IF NOT EXISTS random_data2 (
    id SERIAL PRIMARY KEY,
    random_text1 TEXT,
    random_text2 TEXT,
    random_text3 TEXT,
    random_text4 TEXT,
    random_text5 TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

CREATE TABLE IF NOT EXISTS random_data3 (
    id SERIAL PRIMARY KEY,
    random_text1 TEXT,
    random_text2 TEXT,
    random_text3 TEXT,
    random_text4 TEXT,
    random_text5 TEXT,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);

-- Create a function to populate the tables with random data
CREATE OR REPLACE FUNCTION populate_random_data(batch_size INTEGER)
RETURNS VOID AS $$
BEGIN
    -- Populate random_data1 table
    INSERT INTO random_data1 (random_text1, random_text2, random_text3, random_text4, random_text5)
    SELECT 
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT)
    FROM generate_series(1, batch_size);

    -- Populate random_data2 table
    INSERT INTO random_data2 (random_text1, random_text2, random_text3, random_text4, random_text5)
    SELECT 
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT)
    FROM generate_series(1, batch_size);

    -- Populate random_data3 table
    INSERT INTO random_data3 (random_text1, random_text2, random_text3, random_text4, random_text5)
    SELECT 
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT),
        md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT) || md5(random()::TEXT)
    FROM generate_series(1, batch_size);
END;
$$ LANGUAGE plpgsql;

-- SELECT generate_gigabytes_of_data(5); -- This will generate approximately 5GB of random data

counter=1; while true; do echo "Batch number: $counter"; psql -c "select populate_random_data(10000);"; ((counter++)); done

```

```
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
2024-07-03 21:26:49,641 INFO: Lock owner: db-sgrtest10-instance1-8ccj-0; I am db-sgrtest10-instance1-zj58-0
2024-07-03 21:26:49,641 INFO: bootstrap from leader 'db-sgrtest10-instance1-8ccj-0' in progress
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
/tmp/postgres:5432 - rejecting connections
2024-07-03 21:26:55,674 INFO: Lock owner: db-sgrtest10-instance1-8ccj-0; I am db-sgrtest10-instance1-zj58-0
2024-07-03 21:26:55,674 INFO: Still starting up as a standby.
2024-07-03 21:26:55,677 INFO: Lock owner: db-sgrtest10-instance1-8ccj-0; I am db-sgrtest10-instance1-zj58-0
2024-07-03 21:26:55,677 INFO: establishing a new patroni connection to the postgres cluster
2024-07-03 21:26:56,109 INFO: establishing a new patroni connection to the postgres cluster
2024-07-03 21:26:56,356 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
/tmp/postgres:5432 - accepting connections
2024-07-03 21:26:59,760 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:09,774 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:19,725 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:29,739 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:39,751 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:49,758 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
2024-07-03 21:27:59,764 INFO: no action. I am (db-sgrtest10-instance1-zj58-0), a secondary, and following a leader (db-sgrtest10-instance1-8ccj-0)
```

### Environment

- Platform: Kubernetes/GKE
- Platform Version: 1.27.x
- PGO Image Tag: 5.4.1-0
- Postgres Version: 13
- Storage: GKE SSDs

**Comment by andrewlecuyer** [comment]
Hi @Antaxify!  My assumption here is that the replica is simply replaying all of the WAL that is currently available in the pgBackRest repository at the time the replica is created.  This would mean some of the WAL from the primary instance simply hasn't been pushed and archived yet.

I suggest creating a backup between steps 3 and 4, i.e. [manually create a new backup](https://access.crunchydata.com/documentation/postgres-operator/latest/tutorials/backups-disaster-recovery/backup-management#taking-a-one-off-backup) before creating the replica.  This should ensure the replica is bootstrapped from a backup that includes all of your data.

**Comment by Antaxify** [comment]
Hey @andrewlecuyer!
Thank you for taking the time to answer. I intentionally skip a full backup in between to test exactly this use case. I figured out what's happening in the meantime:

The readiness and liveness probes use the patroni REST endpoint `/readiness` and `/liveness`, which both don't wait for the cluster to be fully synced.

I ended up using `/replica?lag=0MB` in my script to check if the replica is fully synced, which takes nearly 15 minutes in the example above.

Found it in the patroni docs: https://patroni.readthedocs.io/en/latest/rest_api.html

Unfortunately, it seems to be impossible to customize the `readinessProbe` and `livenessProbe` commands, or ideally add a `startupProbe`.
